public/dorequest.php

```diff
@@ -430,7 +432,7 @@ function DoRequestError(string $error, ?int $status = 200, ?string $code = null)
         foreach ($awardableAchievements as $achievement) {
             $unlockAchievementResult = unlockAchievement($targetUser, $achievement->id, $hardcore);
 
-            if (!isset($unlockAchievementResult['Error'])) {
+            if (!isset($unlockAchievementResult['Error']) && $achievement->is_published) {
                 dispatch(new UnlockPlayerAchievementJob($targetUser->id, $achievement->id, $hardcore))
                     ->onQueue('player-achievements');
```

Comment:
I wonder if the logic should be in the `UnlockPlayerAchievement` action? Are there other places that call the action? Will there be?

The job also updates the session, maybe we still want that to happen?

Reply:
Good point. I've removed `$achievement->is_published` from the conditional and updated `UnlockPlayerAchievement` to instead use:

```php
if ($achievement->is_published) {
    $unlock->save();
}
```